### PR TITLE
Show agility level overhead (either as icons or text) in the wilderness

### PR DESCRIPTION
**Pull request by chrisjohnson**
As well as some general player indicator improvements around multiple
conflicting options

I'm not sure if it's appropriate to stuff hiscore checking code in the overlay but it seemed cumbersome to try to do it in the plugin and surface the data to the overlay. LMK if you have other thoughts

### CODE REVIEW

**Comment by Ganom** [comment]
Http calls should never be ran on a loop, check my scouter for an example on how to properly implement look ups on players.

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 1]
```diff
@@ -232,6 +231,43 @@ private void renderPlayerOverlay(Graphics2D graphics, Player actor, Color color)
 					textLocation.getY());
 			}
 		}
+
+		if (plugin.isShowAgilityLevel())
+		{
+			if (!plugin.agilityMap.isEmpty() && plugin.agilityMap.keySet().contains(actor.getName()))
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: remove the `isEmpty()` call

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 2]
```diff
@@ -232,6 +231,43 @@ private void renderPlayerOverlay(Graphics2D graphics, Player actor, Color color)
...
+			{
+				if (plugin.getAgilityFormat().equals(PlayerIndicatorsPlugin.AgilityFormats.ICONS))
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: make this a switch statement or use `==`

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 3]
```diff
@@ -232,6 +231,43 @@ private void renderPlayerOverlay(Graphics2D graphics, Player actor, Color color)
...
+				{
+					int width = graphics.getFontMetrics().stringWidth(name);
+					int height = graphics.getFontMetrics().getHeight();
+					if (plugin.agilityMap.get(actor.getName()) >= 70)
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: `70` should be a config option or a `static final` field

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 4]
```diff
@@ -232,6 +231,43 @@ private void renderPlayerOverlay(Graphics2D graphics, Player actor, Color color)
...
+					{
+						OverlayUtil.renderImageLocation(graphics,
+								new Point(textLocation.getX() + 5 + width,
+									textLocation.getY() - height),
+								ImageUtil.resizeImage(agilityIcon, height, height));
+					}
+					if (plugin.agilityMap.get(actor.getName()) >= 84)
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: `84` should be a config option or a `static final` field

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 5]
```diff
@@ -232,6 +231,43 @@ private void renderPlayerOverlay(Graphics2D graphics, Player actor, Color color)
...
+					{
+						OverlayUtil.renderImageLocation(graphics,
+								new Point(textLocation.getX() + agilityIcon.getWidth() + width,
+									textLocation.getY() - height),
+								ImageUtil.resizeImage(agilityIcon, height, height));
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: caching again

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 6]
```diff
@@ -232,6 +231,43 @@ private void renderPlayerOverlay(Graphics2D graphics, Player actor, Color color)
...
+				if (plugin.getAgilityFormat().equals(PlayerIndicatorsPlugin.AgilityFormats.ICONS))
+				{
+					int width = graphics.getFontMetrics().stringWidth(name);
+					int height = graphics.getFontMetrics().getHeight();
+					if (plugin.agilityMap.get(actor.getName()) >= 70)
+					{
+						OverlayUtil.renderImageLocation(graphics,
+								new Point(textLocation.getX() + 5 + width,
+									textLocation.getY() - height),
+								ImageUtil.resizeImage(agilityIcon, height, height));
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: cache this in a `@Getter private BufferedImage agilityScaled = null;` and set it at the top of `if (plugin.isShowAgilityLevel())` using `if (agilityScaled == null) // resize and save`

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java [hunk 7]
```diff
@@ -232,6 +231,43 @@ private void renderPlayerOverlay(Graphics2D graphics, Player actor, Color color)
...
+					}
+				}
+				else
+				{
+					name += " " + plugin.agilityMap.get(actor.getName());
+
+					int width = graphics.getFontMetrics().stringWidth(name);
+					int height = graphics.getFontMetrics().getHeight();
+					OverlayUtil.renderImageLocation(graphics, new Point(textLocation.getX() + 5 + width,
+																		textLocation.getY() - height),
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: weird indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java [hunk 8]
```diff
@@ -218,6 +240,38 @@ private void onClanMemberLeft(ClanMemberLeft event)
...
+		final Player player = event.getPlayer();
+
+		if (player == client.getLocalPlayer())
+		{
+			return;
+		}
+
+		if (checkWildy() && isShowAgilityLevel())
+		{
+			if (!hiscoreCheckBlacklist.isEmpty() && hiscoreCheckBlacklist.keySet().contains(player.getName()))
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: remove `isEmpty()`
if `hiscoreCheckBlacklist.getOrDefault(player.getName(), 1) < client.getTickCount()`

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java [hunk 9]
```diff
@@ -155,6 +173,9 @@
 	@Getter(AccessLevel.PACKAGE)
 	private boolean unchargedGlory;
 
+	public Map<String, Integer> agilityMap = new HashMap<>();
+	private ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: allocate a higher number of threads, in places like rev caves this cap can easily be exceeded on lower end computers

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java [hunk 10]
```diff
@@ -277,6 +277,30 @@ default boolean showCombatLevel()
...
+	default boolean showAgilityLevel()
+	{
+		return false;
+	}
+
+	@ConfigItem(
+		position = 20,
+		keyName = "agilityFormat",
+		name = "Format",
+		description = "Whether to show the agility level as text, or as icons (1 skull >= 70, 2 skulls >= 84).",
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java]: add configs for the `70` and `84`

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java [hunk 11]
```diff
@@ -218,6 +240,38 @@ private void onClanMemberLeft(ClanMemberLeft event)
...
+			if (!hiscoreCheckBlacklist.isEmpty() && hiscoreCheckBlacklist.keySet().contains(player.getName()))
+			{
+				return;
+			}
+
+			executorService.submit(() ->
+			{
+				try
+				{
+					hiscoreCheckBlacklist.put(player.getName(), client.getTickCount() + 6000); // TODO : Config? Handle clearing expired entries?
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: move before the `try` or into a `finally` after the catch
use a `private static final ONE_HOUR = 6000;`

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java [hunk 12]
```diff
@@ -218,6 +240,38 @@ private void onClanMemberLeft(ClanMemberLeft event)
...
+					SingleHiscoreSkillResult result = HISCORE_CLIENT.lookup(player.getName(), HiscoreSkill.AGILITY);
+					Integer level = result.getSkill().getLevel();
+					agilityMap.put(player.getName(), level);
+				}
+				catch (IOException ex)
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: print the exception

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java [hunk 13]
```diff
@@ -277,6 +277,30 @@ default boolean showCombatLevel()
...
+		group = "Target Indicator"
+	)
+	default PlayerIndicatorsPlugin.AgilityFormats agilityFormat()
+	{
+		return PlayerIndicatorsPlugin.AgilityFormats.ICONS;
```
Comment (ThatGamerBlue) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java]: default to `TEXT`

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java [hunk 14]
```diff
@@ -399,11 +453,22 @@ else if (player.getOverheadIcon().equals(HeadIcon.SMITE))
 		}
 	}
 
+	private boolean checkWildy()
+	{
+		return client.getVar(Varbits.IN_WILDERNESS) == 1 || WorldType.isPvpWorld(client.getWorldType());
```
Comment (Kyleeld) [inline runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: WorldType.isAllPvpWorld(client.getWorldType());

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java (hunk 7 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: How would you re-indent? textLocation is an argument to Point so it seems right to indent further here, plus my editor agrees. But I'll take a suggestion

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java (hunk 2 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: Done

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java (hunk 3 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: Done

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java (hunk 4 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: Done

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java (hunk 8 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: I switched to using HiscoreManager which already does the caching for me

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java (hunk 10 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java]: Done

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java (hunk 9 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: Got rid of my own executor

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java (hunk 13 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsConfig.java]: Done

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java (hunk 14 above)
Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsPlugin.java]: Done

File: runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java (hunk 7 above)
Comment (ThatGamerBlue) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: I'd do it like this, like it is just above https://i.imgur.com/3wm98Nv.png

Comment (chrisjohnson) [reply runelite-client/src/main/java/net/runelite/client/plugins/playerindicators/PlayerIndicatorsOverlay.java]: Better?

**Pull request merged**